components/react/hooks/src/useRequest/index.js

```diff
@@ -0,0 +1,179 @@
+import {useCallback, useEffect, useReducer, useRef} from 'react'
+
+export const STATUSES = {
+  IDLE: 'idle',
+  LOADING: 'loading',
+  ERROR: 'error',
+  SUCCESS: 'success'
+}
+
+const ACTIONS = {
+  FETCHING: 'fetching',
+  LOADING: 'loading',
+  ERROR: 'error',
+  SUCCESS: 'success',
+  REPLACE: 'replace',
+  RESET: 'reset'
+}
+
+const initialState = {
+  status: STATUSES.LOADING,
+  data: null,
+  error: null,
+  isFetching: false
+}
+
+const getInitialState = ({initialData, initialStatus}) => {
+  const data = typeof initialData === 'function' ? initialData() : initialData
+  const status = data !== undefined ? STATUSES.SUCCESS : initialStatus || initialState.status
+
+  return {
+    status,
+    data,
+    error: null
+  }
+}
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case ACTIONS.FETCHING:
+      return {
+        ...state,
+        isFetching: true
+      }
+    case ACTIONS.LOADING:
+      return {
+        ...initialState,
+        isFetching: true,
+        data: state.data
+      }
+    case ACTIONS.ERROR:
+      return {
+        ...initialState,
+        status: STATUSES.ERROR,
+        error: action.payload.error,
+        data: state.data
+      }
+    case ACTIONS.SUCCESS:
+      return {
+        ...initialState,
+        status: STATUSES.SUCCESS,
+        data: action.payload.data
+      }
+    case ACTIONS.REPLACE:
+      return {
+        ...initialState,
+        status: STATUSES.SUCCESS,
+        data: action.payload.isHard ? action.payload.data : {...state.data, ...action.payload.data}
+      }
+    case ACTIONS.RESET:
+      return {
+        ...initialState,
+        status: action.payload.initialStatus
+      }
+    default:
+      return state
+  }
+}
+
+const useRequest = (request, {initialStatus, initialData, onSuccess, onError} = {}) => {
+  const [{status, data, ...others}, dispatch] = useReducer(reducer, getInitialState({initialStatus, initialData}))
+  const requestRef = useRef(request)
+  const onSuccessRef = useRef(onSuccess)
+  const onErrorRef = useRef(onError)
+  const isMountedRef = useRef(false)
+
+  const handleError = error => {
+    dispatch({type: ACTIONS.ERROR, payload: {error}})
+
+    if (onErrorRef.current) {
+      onErrorRef.current(error)
+    }
+  }
+
+  const fetch = useCallback(async (...args) => {
+    try {
+      const [error, data] = await requestRef.current(...args)
+
+      if (!isMountedRef.current) {
+        return
+      }
+
+      if (error) {
+        return handleError(error)
+      }
+
+      dispatch({type: ACTIONS.SUCCESS, payload: {data}})
+
+      if (onSuccessRef.current) {
+        onSuccessRef.current(data)
+      }
+    } catch (error) {
+      handleError(error)
+    }
+  }, [])
+
+  const refresh = useCallback(
```

Comment:
If `fetch` is already memorized, the use of `useCallback` here might be a downgrade instead of an upgrade in performance. The same for the other functions that only depend on the `fetch`